### DISCUSSION TITLE
Installing insiders via poetry

### DISCUSSION THREAD

**Discussion by madeinoz67** [squidfunk/mkdocs-material#5076] [Q&A]
I want to install insider via poetry however it saves the token in the pyproject.toml regardless of using an ENV variable for the token.

how are others doing installing insiders via poetry?

**Comment by gordonhart** [comment]
I saw two options when setting up insiders for [kolena](https://github.com/kolenaIO/kolena):

1. Add the `mkdocs-material-insiders` repo (or private fork) as a git dependency to `pyproject.toml`
    - Only users with access to the insiders repo will be able to use the entire Poetry project (not great!)
 
2. Add the publicly-accessible `mkdocs-material` dependency to `pyproject.toml` and use an out-of-band installation script (example: [`docs/setup_insiders.sh`](https://github.com/kolenaIO/kolena/blob/trunk/docs/setup_insiders.sh)) to install `mkdocs-material-insiders` on top of the Poetry environment, using `poetry run pip install ...` with a git dependency.
    - This means users without access to the insiders repo can still install the project and build the docs (sans insiders features), and developers with access to the insiders repo can run this script to install insiders and build the docs with insiders features.
    - Note that Poetry actions such as `poetry add` or `poetry install` will reinstall the base `mkdocs-material` dependency, meaning the `./docs/setup_insiders.sh` script needs to be re-run to install the insiders deps on top of the environment.

We went with route (2) and it's been working pretty well for us. This pattern also works for the insiders fork of the wonderful [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) package.

Curious to hear if others have found a better route.

**Reply by SongshGeo** [reply]
Hi, sorry about the necrobumping. 😿

I managed to use your second suggestion effectively but with a token from my private account.
According to the `setup_insiders.sh` snippet you open-sourced, it seems like the `mkdocs-material-insider` fork is accessible in Kolena's organization account. I'm curious about how to do so, accessing the private organization repo through poetry. Do you have any advice on this?

**Comment by pawamoy** [comment]
When I started playing with Material for MkDocs Insiders, I wanted to make its installation transparent. I wanted maintainers (me) to be able to install the Insiders version, while still allowing contributors without access to install the public version.

```toml
[project]
dependencies = [
    "mkdocs-material>=9",  # match insiders if you have access, match public version otherwise
]
```

Well, I found a way, using a local, self-hosted PyPI-like index :slightly_smiling_face: 

Long story short: I have released a project called **PyPI Insiders** (as part of [my own Insiders program](https://pawamoy.github.io/insiders/), very much inspired by @squidfunk's) that manages Insiders repositories for you:

- configure which insiders repositories to watch (such as squidfunk/mkdocs-material-insiders for example)
- serve a local PyPI server (or not, you can upload to any other PyPI-like index)
- PyPI Insiders will periodically (or whenever you manually instruct it to) pull new commits within your configured repositories, build artifacts for the latest tags, and upload the distributions to your self-hosted PyPI index
- tools like pip, PDM or Poetry can then be configured to use this self-hosted index to search and install packages

Check it out here: https://pawamoy.github.io/pypi-insiders/

I tried to make the installation as smooth as possible, though it is not battle-tested, and docs are still a bit lacking, so if you try it and have some trouble, don't hesitate to open issues (on the public bugtracker: https://github.com/pawamoy/pypi-insiders/issues)!
